Content libraries MVP

bradenmacdonald opened the pull request:
**Description**: This is our minimum viable implementation of content libraries ("Problem Banks") ([SOL-11](https://openedx.atlassian.net/browse/SOL-11)). This PR will be used for QA and bug fixing before we merge this feature branch to master.

**Sandbox URL**: CMS http://content-libraries.sandbox.opencraft.com:18010/ - LMS http://content-libraries.sandbox.opencraft.com/

**Partner information**: 3rd party-hosted open edX instance, for an edX solutions client.

**Dependencies**: Already merged: Support for Content libraries in the modulestore (#6033) and in opaque keys (https://github.com/edx/opaque-keys/pull/46)

**Constituent PRs**: All the code on this branch has already gone through a code review, except for 76d3f359694841e9327f4688abb0144a2c9f8f72 and any future bug fixes we make once all the constituent PRs have been merged:

- [x] Studio support for creating and editing libraries #6046 (SOL-1, SOL-2, SOL-3)
- [x] LibraryContent - Display content from a library in a course #6155 (SOL-5, SOL-6, SOL-7, SOL-8, SOL-117)
- [x] Pagination for Content libraries  #6163 (SOL-4)
- [x] Content library permissions & collaboration #6283 (SOL-81)
- [x] Filter library content by CAPA problem type #6346 (SOL 122)
- [x] Content Libraries UI Updates #6388 (SOL-80)
- [x] Content libraries settings overrides #6399 (SOL-46)
- [x] Content libraries analytics enhancements #6492 (SOL-121) - depends on the one above
- [x] Sidebar text for Content Library page in Studio #6426
- [x] Improved test coverage #6499

CC @antoviaque @Kelketek @e-kolpakov @mtyaka

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?number=6459&repo=edx%2Fedx-platform

antoviaque commented:
@bradenmacdonald The jenkins test run doesn't seem to have been automatically triggered on this?

@jzoldak Any idea why this would not happen? Is there a way for us to trigger them manually when this happens? Usually we amend the last commit to attempt to retrigger a jenkins run, but on a feature branch with PRs that depend on it, that would require to rebase all dependent PRs afterwards, which wouldn't be practical.

antoviaque commented:
FYI QA tests are all green: https://openedx.atlassian.net/wiki/display/SOL/Content+Libraries+-+Testing+Results

jzoldak commented:
@antoviaque  @bradenmacdonald 
We're in the middle of switching to a new jenkins infrastructure. Here are URLs to trigger jobs manually on the [old jenkins server](https://jenkins.testeng.edx.org/job/edx-all-tests-manual-pr/) and the [new one](https://build.testeng.edx.org/job/edx-platform-all-tests/build).

bradenmacdonald commented:
@antoviaque @chrisndodge @cahrens @explorerleslie @Kelketek @e-kolpakov @mtyaka This MVP PR for content libraries is just about ready to go! Can one of you at edX please just review the two tiny "catch up" commits I just pushed: 6ea5a8a595d993795178ca6b9ad59af8edd1efd9 (Fix hard coded studio name) and 012e6e6fd336d04693551e6c73185eed2de88b4f (Update font awesome icons)? Everything else has been reviewed. Then I believe this can be merged :)

e-kolpakov commented:
@antoviaque @chrisndodge @cahrens @explorerleslie @Kelketek @bradenmacdonald @mtyaka We have found a bug in library container JS - actual DOM structure differs from what it expects, so under certain circumstances new XBlocks appear *inside* their sibling blocks' wrappers. As a result, if "host" block is removed, all of it's "guests" are removed as well.

This is a client-side only issue (does not persist between page reloads). Fix and a more verbose description is at [this PR](https://github.com/edx/edx-platform/pull/6580).

I believe we need that PR merged before proceeding with `content-libraries` -> `master` merge.

antoviaque commented:
@e-kolpakov To make it easier to review everything at once for the MVP merge, I've merged that bugfix here too. So the commits to review to be able to merge the current MVP PR to master are: https://github.com/edx/edx-platform/commit/6ea5a8a595d993795178ca6b9ad59af8edd1efd9 https://github.com/edx/edx-platform/commit/012e6e6fd336d04693551e6c73185eed2de88b4f 
https://github.com/edx/edx-platform/commit/9f6cf1530eee3ca971f73e485999a2d4650c2d9c

chrisndodge commented:
@bradenmacdonald any chance to write - within a reasonable level of effort - a quick automated test regarding the fix at 9f6cf1530eee3ca971f73e485999a2d4650c2d9c?

chrisndodge commented:
Other than my suggestion re: adding a unit test that would have caught the bug that was fixed at 9f6cf1530eee3ca971f73e485999a2d4650c2d9c, looks good to me.

Congratulations!

e-kolpakov commented:
@chrisndodge I'm writing autotest right now.

antoviaque commented:
@chrisndodge Thank you for having a look so quickly! You even arrived before @e-kolpakov had had time to add the test : ) It's now done, and I added it to the current PR: https://github.com/edx/edx-platform/commit/720834c450af773daebc060b3df3d894c88d0b9c

chrisndodge commented:
Thanks for the new acceptance test which simulates that previous error condition.

LGTM. +1

cahrens commented:
:+1: I have a requested clarification on the test case, but otherwise LGTM. Congrats on finishing up all this work!

e-kolpakov commented:
@cahrens yes the text is there only to assert on it later. The approach is reused from other test a couple of lines above.

@bradenmacdonald I'm about logging off for today, so could you please polish the tests if needed?

bradenmacdonald commented:
@cahrens I have simplified the bok choy test as you requested: https://github.com/edx/edx-platform/commit/bce8ee69#diff-1 . I also squashed the various commits for this last bugfix. Hope that's good? I'd like to merge this once the build is green.

cahrens commented:
Looks good! :+1:

bradenmacdonald commented:
Content Libraries / Problem Banks MVP is now merged. Congratulations and thanks to everyone who contributed to it!

chrisndodge commented:
@bradenmacdonald congratulations and thanks for all the hard work!

I just realized that neither @cahrens nor I asked you to squash your commits before merging at the end of the review, which we typically do. Typically, we prefer a single commit before merging.

I'm not sure if I'd suggest reverting this and squashing and re-applying. @cahrens @smagoun what do you think?

bradenmacdonald commented:
@chrisndodge Ah! Sorry :-/ I didn't realize that was the case. (I personally find that having more of the commit history helps future developers with understanding the code changes that were made in this PR.) I can definitely do that if you'd like, but it will involve resetting master back and thus a force push to master.

But anyhow just let me know - I'm happy to squash this whole thing down to a single commit, re-run the build, and re-merge if you'd like me to.

chrisndodge commented:
Yea, reverting, squashing, re-merging might be more hassle than it is worth at this point in time. I wouldn't do anything right now, unless anyone else has a major opinion on this.

Going forward, we'll try to remember to remind all contributors to squash commits after the review is done.

cahrens commented:
Yes, @andy-armstrong is now struggling through some rebasing. I'm sorry I forgot to check if commits were squashed. 

Note that we don't necessarily require a single commit per PR, but commits should be logical chunks of work (for instance, a commit per PR merged into the feature branch is fine, if the commits represent stories). Commits like "pylint fixes" or "Addressed nits" aren't helpful and increase the merge conflicts.

bradenmacdonald commented:
Ok, let me know if any action is required from me. I agree that ~one commit per PR merged to this branch would be nice.

cahrens commented:
It would be nice if ENABLE_CONTENT_LIBRARIES existed in the common envs.py file (with default value of False) so it was easier to turn on in sandboxes. I think @andy-armstrong is going to do that as part of cleaning up the conflicts we are having with setting visibility on an xblock on the container page.

@bradenmacdonald and @marcotuts does this look right to you when there are no content libraries?

![image](https://cloud.githubusercontent.com/assets/484484/5729704/5fb4e45c-9b42-11e4-9860-d8450772fd63.png)

cahrens commented:
@bradenmacdonald and @marcotuts Note also that error messaging should be updated. It isn't possible to have both a library and a course with the same org and course number.

![image](https://cloud.githubusercontent.com/assets/484484/5729959/7886b7ec-9b44-11e4-90e6-60971fd14242.png)

e-kolpakov commented:
@jzoldak it appears build triggers requires some sort of special permission. At least it doesn't work for me: "e-kolpakov is missing the Task/Build permission", despite I've granted all permissions it asked for. Is that supposed to be used for core devs only, or access is granted on-demand?

jzoldak commented:
@e-kolpakov Requirements are that you need to be a [member of the edx org on GitHub](https://github.com/orgs/edx/people) and you need to make that membership public.
@antoviaque can you help him with this?

antoviaque commented:
@jzoldak Thanks for the info - yup, I'll see if we can get Eugeny in that group, perfect.
